File: 909/c/main.c
#include <stdbool.h>
#include <string.h>

int lookup_tile(int** board, int n, int tile) {
    int cap = n - 1;
    int index = tile - 1;
    int column = index % n;
    int row = cap - ( index / n );
    if ((cap-row) % 2) {
        column = cap - column;
    }
    int dest = board[row][column];
    if (dest == -1) {
        return tile;
    } else {
        return dest;
    }
}

int snakesAndLadders(int** board, int boardSize, int* boardColSize) {
    int n = boardSize;
    int end = n * n;
    int turn = 1;
    int next_turn_len = 1;
    int next_turn[end];
    next_turn[0] = 1;
    int next_turn_alt[end];
    bool seen[end];
    memset(seen, 0, sizeof seen);

    int* nt = next_turn;
    int* nt_ = next_turn_alt;

    while (next_turn_len > 0) {
        int nt_len = 0;
        for (int i = 0; i < next_turn_len; i++) {
            int tile = nt[i];
            for (int next = tile + 1; next <= tile + 6; next++) {
                if (next == end) { return turn; }
                if (next > end) { continue; }
                int dest = lookup_tile(board, n, next);
                if (dest == end) { return turn; }
                if (!seen[dest]) {
                    seen[dest] = true;
                    nt_[nt_len++] = dest;
                }
            }
        }

        // swap array pointers
        int* shim = nt;
        nt = nt_;
        nt_ = shim;
        next_turn_len = nt_len;

        turn++;
    }
    return -1;
}

File: 909/python/main.py
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)
        end = n * n
        turn = 1
        seen = [False] * end
        next_turn = [1]
        while len(next_turn) > 0:
            for tile in next_turn[:]:
                next_turn.remove(tile)
                for next in range(tile + 1, tile + 7):
                    if next == end:
                        return turn
                    if next > end:
                        continue
                    dest = lookup_tile(board, n, next)
                    if dest == end:
                        return turn
                    if not seen[dest]:
                        seen[dest] = True
                        next_turn.append(dest)
            turn += 1
        return -1

def lookup_tile(board, n, tile):
    cap = n - 1
    index = tile - 1
    column = index % n
    row = cap - ( index // n )
    if (cap - row) % 2:
        column = cap - column # reverse
    dest = board[row][column]
    if dest != -1:
        return dest
    else:
        return tile

